nengo_loihi/hardware/nxsdk_objects.py

```diff
@@ -260,20 +260,30 @@ def set_axons(self, board, nxsdk_board, spike_input):
         for axon in spike_input.axons:
             axon_type = axon.pop_type
             assert axon_type in (0, 32), "Only discrete and pop32 supported"
-            tchip_idx, tcore_idx, tsyn_ids = board.find_synapse(axon.target)
+            synapse = axon.target
+            tchip_idx, tcore_idx, tsyn_ids = board.find_synapse(synapse)
             tchip = d_get(nxsdk_board, b"bjJDaGlwcw==")[tchip_idx]
             tcore = d_get(tchip, b"bjJDb3Jlcw==")[tcore_idx]
             spikes = axon.map_spikes(input_idxs)
             for input_idx, spike in zip(input_idxs, spikes):
-                if spike is not None:
-                    taxon_idx = int(spike.axon_id)
-                    taxon_id = int(tsyn_ids[taxon_idx])
-                    self.axon_map.setdefault(input_idx, []).append(
-                        np.array(
-                            (-1, axon_type, tchip.id, tcore.id, taxon_id, spike.atom),
-                            dtype=self.spike_dtype,
-                        )
+                if spike is None:
+                    # This should not happen, because input slices happen when connecting
+                    # into the spike input. However, if this arises we just skip it.
```

Comment:
In `Axon.compartment_map`, we allow a compartment to have an entry of -1 to indicate that that compartment doesn't have any output associated with that `Axon` object. If you call `Axon.map_spikes` with that compartment index, you'll get a `None` spike back. So if we encounter `spike is None` here, it just means that that compartment isn't making use of this `Axon`, so we skip it.

However, when I was testing, I realized that because the way we do our slices if e.g. you've got a sliced pre object off-chip going to a post object on-chip, the slice happens before the `SpikeInput`, so the axon coming from the `SpikeInput` never has compartments with a negative entry in `compartment_map`.

I guess we could test this by manually setting up a `SpikeInput` with a custom `Axon` that does have negative entries for some compartments. It is a feature that the builder could theoretically use, it's just that as it currently happens, it doesn't. So maybe we just need more of a unit test rather than an integration test for this function.

Reply:
Why do we add the compartment to the map if there are no axons associated with it?

Reply:
Because the map is just a list/array, so it needs an index for each compartment. We could implement it as a dict, but I think I've been making use of the ability to get values for multiple indices simultaneously with an array in some places.

Reply:
I looked into this a fair bit but a lot of the classes in `block.py` do similar things as `Axon` (delayed initialization that is not encapsulated in the class, allowing things like mapping spikes that don't exist, etc) and solving all those issues would take too much time, so I made #270 with the changes I made in case we want to deal with that refactoring later.